**bold text**# HW 2 Part 2: Association Rule Mining on Bike Rentals

The original dataset is available here: https://s3.amazonaws.com/baywheels-data/index.html 
We have a filtered subset of this dataset for this homework. 

There are two datasets - one for the month of February and one for the month of March 2022. Each dataset contains "buckets" of bike rentals. Each row corresponds to an hour in the month. For example, row 1 of the February dataset corresponds to bike rentals starting on February 1st 12AM-1AM. The columns where the value is "True" correspond to the bike rental starting locations in the San Francisco Bay Area. 

If your SJSU ID ends in an odd number, pick the February dataset. Otherwise, pick the March dataset.

Perform the following tasks on the provided bike rental dataset. Complete all the LP questions to receive a "low pass" grade on the homework. Complete all the questions (both LP and HP) to receive a "high pass" grade on the homework. 

Note that if you are unable to complete any of the LP questions satisfactorily, you will receive a grade of "revision required". You can revise and resubmit your work in exchange for a token. Please review the syllabus for more information on specifications grading.

**VERY IMPORTANT**: Include **ALL** the references you used for this assignment, including names of classmates you discuss with. Failure to cite your sources counts as an act of academic dishonesty and will be taken seriously without zero tolerance. You will automatically receive a “fail” grade in the homework and further serious penalties may be imposed.

NOTE: You can look for help on the Internet but refrain from referencing too much. Please cite all your sources in your submission. 
When you submit your assignment, you automatically agree to the following statement. If you do not agree, it is your responsibility to provide the reason.

“*I affirm that I have neither given nor received unauthorized help in completing this homework. I am not aware of others receiving such help. I have cited all the sources in the solution file.*”

In [ ]:
import pandas as pd
## Your code goes here. Import the csv into a pandas dataframe here
data_frame_march=pd.read_csv("202203-baywheels-dataset-basket.csv")


[LP] Perform Association Rule Mining using the apriori algorithm.
1. If you are using the February dataset, run the algorithm with a support above 0.55 and confidence above 0.88.
2. If you are using the March dataset, run the algorithm with a support above 0.6 and confidence above 0.8.

The library you can use is mlxtend. The imports are included below. 



In [ ]:
## Your code for LP goes here.
import mlxtend
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

## Selecting support value greater than 0.6
march_frequent_itemsets=apriori(data_frame_march,min_support=0.6,use_colnames=True)
march_frequent_itemsets



support                                           itemsets
0   0.634771                           (17th St at Valencia St)
1   0.668464                               (Berry St at 4th St)
2   0.615903  (Civic Center/UN Plaza BART Station (Market St...
3   0.712938                             (Market St at 10th St)
4   0.668464                          (Market St at Steuart St)
5   0.610512  (Montgomery St BART Station (Market St at 2nd ...
6   0.625337                               (Polk St at Post St)
7   0.614555                             (Post St at Kearny St)
8   0.676550     (Powell St BART Station (Market St at 4th St))
9   0.648248     (Powell St BART Station (Market St at 5th St))
10  0.603774  (Market St at 10th St, Powell St BART Station ...

In [ ]:
## Selecting rules having confidence greater than 0.8
march_result=association_rules(march_frequent_itemsets,metric="confidence",min_threshold=0.8)
march_result


antecedents  \
0                          (Market St at 10th St)   
1  (Powell St BART Station (Market St at 4th St))   

                                      consequents  antecedent support  \
0  (Powell St BART Station (Market St at 4th St))            0.712938   
1                          (Market St at 10th St)            0.676550   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.676550  0.603774    0.846881  1.251764  0.121435    2.112409  
1            0.712938  0.603774    0.892430  1.251764  0.121435    2.668613

From the above results we can observe the following insights:
1)For the association rule "Market St at 10th St", "Powell St BART Station (Market St at 4th St)" the support is 60% and the confidence is 84%.

2)For the association rule "Market St at 10th St", "Powell St BART Station (Market St at 4th St)" the support is 60% and the confidence is 89%.


[HP] Perform association rule mining on the other dataset as well. Explain briefly what the association rules mean. 

In [ ]:
## Your code for HP goes here. 
import pandas as pd
## Your code goes here. Import the csv into a pandas dataframe here
data_frame_feb=pd.read_csv("202202-baywheels-dataset-basket.csv")

import mlxtend
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

## Selecting support value greater than 0.55
feb_frequent_itemsets=apriori(data_frame_feb,min_support=0.55,use_colnames=True)
feb_frequent_itemsets



support                                           itemsets
0   0.617560                           (17th St at Valencia St)
1   0.590774                               (Berry St at 4th St)
2   0.580357  (Civic Center/UN Plaza BART Station (Market St...
3   0.559524                            (Fell St at Stanyan St)
4   0.553571                              (Folsom St at 5th St)
5   0.553571                            (Laguna St at Hayes St)
6   0.699405                             (Market St at 10th St)
7   0.662202                          (Market St at Steuart St)
8   0.587798                             (Mission Dolores Park)
9   0.577381                               (Mission Playground)
10  0.561012                           (Pierce St at Haight St)
11  0.555060                               (Polk St at Post St)
12  0.647321     (Powell St BART Station (Market St at 4th St))
13  0.613095     (Powell St BART Station (Market St at 5th St))
14  0.635417  (San Francisco Caltrain Station (King St at 4t...
15  0.561012                            (Sanchez St at 15th St)
16  0.569940              (Union Square (Powell St at Post St))
17  0.583333    (Market St at 10th St, Market St at Steuart St)
18  0.574405  (Market St at 10th St, Powell St BART Station ...
19  0.553571  (Market St at 10th St, Powell St BART Station ...
20  0.563988  (Market St at 10th St, San Francisco Caltrain ...

In [ ]:
## Selecting rules having confidence greater than 0.88
feb_result=association_rules(feb_frequent_itemsets,metric="confidence",min_threshold=0.88)
feb_result


antecedents             consequents  \
0                          (Market St at Steuart St)  (Market St at 10th St)   
1     (Powell St BART Station (Market St at 4th St))  (Market St at 10th St)   
2     (Powell St BART Station (Market St at 5th St))  (Market St at 10th St)   
3  (San Francisco Caltrain Station (King St at 4t...  (Market St at 10th St)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.662202            0.699405  0.583333    0.880899  1.259498   
1            0.647321            0.699405  0.574405    0.887356  1.268731   
2            0.613095            0.699405  0.553571    0.902913  1.290973   
3            0.635417            0.699405  0.563988    0.887588  1.269062   

   leverage  conviction  
0  0.120186    2.523866  
1  0.121665    2.668550  
2  0.124770    3.096131  
3  0.119575    2.674045

< Your answer for HP goes here >

Association Rule:
1) Association rules can be compared to simple if/then statements that helps in finding relations among seemingly independent data lables.
2) Association rule works on non-categorical data sets.
3) Association rule has 2 parts, antecedent and consequent.
4) Anticedent is something occured in the data where consequent is found in combination of Anticedent

On basis of following paramaters, the above relation ships are observed.
i) Support: It defines the number of times if/then relation appears in the database.
ii) Confidence: Confidence tells the number of times these relations have found to be true.

Hence, association rule mining tries to observe the rules that govern how or why such products/items are frequently bought together.

Applications: Market Basket Analysis, Medical Diagnosis, Censes Data etc.


Insights from the data:


#From the above results we can derive following insights:
1) For the association rule "(Market St at Steuart St)", "Powell St BART Station (Market St at 10th St)" the support is 58% and the confidence is 88%.

2)For the association rule "(Powell St BART Station (Market St at 4th St)", "(Market St at 10th St)" the support is 57% and the confidence is 88%.

3)For the association rule "(Powell St BART Station (Market St at 5th St)", "(Market St at 10th St)" the support is 55% and the confidence is 90%.

4)For the association rule "(San Francisco Caltrain Station (King St at 4)", "(Market St at 10th St)" the support is 56% and the confidence is 88%.

# References
Include ALL your references here. 

1) https://www.kaggle.com/code/evrenermis/association-rule-based-learning-explained/notebook

2) http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/

3) Discussed with Teja Ramisetty
# What to turn in:
1. The ipynb solution file, which includes the references
2. If you use Colab or GitHub for version control, please share a link to your notebook or GitHub repository

